File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.java
class Solution {
    
    public int furthestBuilding(int[] heights, int bricks, int ladders) {
    
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for(int i = 1; i < heights.length; i++) {
        
            int diff = heights[i] - heights[i-1];
            
            if(diff > 0) {
            
                if(pq.size() < ladders)                     pq.offer(diff);

                else {
                
                    int br = diff;
                    // Optimize previous ladder use
                    
                    if(pq.size() > 0 && pq.peek() < diff) {
                        br = pq.remove();
                        pq.offer(diff);
                    }
                    
                    if(bricks-br >= 0)                      bricks -=br;
                    else                                    return i-1;
                }
            }
        }
        
        return heights.length - 1;
    }
    
}

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.cpp
class Solution {
public:
    
    // Using Min Heap
    
    int furthestBuilding(vector<int>& h, int bricks, int ladders) {
    
        int n = h.size();
        
        // Using Min Heap
        priority_queue<int, vector<int>, greater<int>> pq;
        
        int i = 0;
        
        while(i < n - 1 && pq.size() < ladders) {
            
            int jumpHeight = h[i + 1] - h[i];
            
            if(jumpHeight > 0)                              pq.push(jumpHeight);
            i++;
        }
        
        while(i < n - 1) {
            
            int jumpHeight = h[i + 1] - h[i];
            
            if(jumpHeight > 0) {
               
                if(pq.empty() == false && jumpHeight > pq.top()) {
                
                    // We here found a larger current jump, so use ladder for this
                    bricks -= pq.top();
                    pq.pop();

                    pq.push(jumpHeight);
                }

                else                                        bricks -= jumpHeight;
            }
            
            if(bricks < 0)                                  return i;
            i++;
        }
        
        return i;
    }
    
};
